compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Fetch.hs

```diff
@@ -0,0 +1,112 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.Compiler.Fetch (
+  LedgerArgs(..), runWithLedgerArgs,
+  createDarFile,
+  fetchDar
+  ) where
+
+import Control.Lens (toListOf)
+import Data.List.Extra (nubSort)
+import Data.String (fromString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as TL
+
+import DA.Daml.Compiler.Dar (PackageSdkVersion(..),createDarFile,createArchive)
+import qualified DA.Daml.LF.Ast as LF
+import qualified DA.Daml.LF.Ast.Optics as LF (packageRefs)
+import qualified DA.Daml.LF.Proto3.Archive as LFArchive
+import qualified DA.Ledger as L
+import qualified SdkVersion
+
+data LedgerArgs = LedgerArgs
+  { host :: String
+  , port :: Int
+  , tokM :: Maybe L.Token
+  , sslConfigM :: Maybe L.ClientSSLConfig
+  }
+
+instance Show LedgerArgs where
+  show LedgerArgs{host,port} = host <> ":" <> show port
+
+-- | Reconstruct a DAR file by downloading packages from a ledger. Returns how many packages fetched.
+fetchDar :: LedgerArgs -> LF.PackageId -> FilePath -> IO Int
+fetchDar ledgerArgs rootPid saveAs = do
+  xs <- downloadAllReachablePackages ledgerArgs rootPid
+  [pkg] <- pure [ pkg | (pid,pkg) <- xs, pid == rootPid ]
+  let (dalf,pkgId) = LFArchive.encodeArchiveAndHash pkg
+  let dalfDependencies :: [(T.Text,BS.ByteString,LF.PackageId)] =
+        [ (txt,bs,pkgId)
+        | (pid,pkg) <- xs, pid /= rootPid
+        , let txt = recoverPackageName pkg ("dep",pid)
+        , let (bsl,pkgId) = LFArchive.encodeArchiveAndHash pkg
+        , let bs = BSL.toStrict bsl
+        ]
+  let (pName,pVersion) = do
```

Comment:
:+1: Thanks!

Reply:
woo hoo :)

Reply:
Now, just hope I'll finally get green on on the CI...